02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md

```diff
@@ -170,26 +170,285 @@ Here's what has been added to the above since then:
+
+The `0.12.0` was a significant achievement but one that came with a lot of breaking changes. Phil responded to it in this way:
+
+> [The PureScript by Example book] seems to take longer [to update] with each major compiler update, and honestly, I’m not motivated to update it again without some sort of guarantee that I won’t have to redo everything again in a few months....
+> I would like to write a lot more documentation for PureScript, but I hate the idea that it will become out of date quickly....
+> So without some sort of plan for the future of the language, and some idea of what changes are coming (and just as importantly, which aren’t), I’m not likely to write anything, and I wouldn’t be surprised if others avoided writing for the same reason.
+>
+> ~ Phil Freeman [The State of Things, point 4](https://discourse.purescript.org/t/the-state-of-things/282)
+
+Gary later pointed out that releases with "breaking changes" were occuring less frequently ([source](https://docs.google.com/spreadsheets/d/1MO8siLLSOpkKHJ_eBNRJbipBquFgqlaSS-QSnALDSo8/edit#gid=0)). Still, keep in mind that Phil was likely updating the book each time such a release was made.
+
+(Related interpretation sections: [Breaking Changes Render Documentation Outdated](https://github.com/chexxor/purescript-documentation-discussion/blob/master/01-Sources-and-Interpretation/All-Interpretations.md#breaking-changes-renders-documentation-outdated))
+
+### There Is No Roadmap That Coordinates Efforts
+
+Defining a roadmap is hard. It's easy to...
+- ...disagree on which goals should be pursued and which to ignore
+- ...disagree on how to word such goals
+- ...do nothing and let someone else contribute
+
+Creating a "good" roadmap takes significant time and energy.
+
+PureScript does not currently have such a roadmap. Rather, core contributors seem to have a general direction they are pursuing (see next point).
+
+Therefore:
+- Those who could help do not know where help is needed
+- Those who would help do not know how they can help
+
+#### Avoiding a `v1.0` PureScript language release
+
+When people announce "Language X is now `1.0`!", it tends to draw a lot of focus and a lot of traffic. People probably think, "Wow! It's now stable enough to be used to write all my programs."
+
+In short, one core contributor believes that many would perceive a `v1.0` release as a `v1.0` ecosystem release. The language could be considered 'good enough' for a `v1.0`. However, the ecosystem is definitely not.
+
+What good is a `v1.0` language, if
+- common libraries haven't stabilized yet or are non-existant?
+- the ecosystem is incoherent in a number of ways?
+- the dependency managers are still unfriendly to users?
+- the IDE support is still lacking?
+
+Thus, core contributors might avoid defining a roadmap to prevent people from having a `v1.0` ecosystem connotation.
```

Comment:
I think having a defined roadmap can be done *before* a `v1.0` release. I imagine a roadmap would specify the features or PRs required before that release will be considered. If the roadmap is for the language only, then I think it would be hard to have the `v1.0` plan depend on other aspects of the language ecosystem. If it's an *ecosystem* roadmap, however, then it's more natural to define that.

I wonder if it's possible to have the PS language maintainers maintain two different roadmaps? If not, then the PS language maintainers would maintain a PS language roadmap, and a different place (I don't know who or where) would maintain a roadmap and version of the ecosystem.

Reply:
> I think having a defined roadmap can be done before a v1.0 release.

I agree. I think having a general roadmap of "what we plan to accomplish this year, regardless of whether that gets us a v1.0 or not" is what should be done. Having a `v1.0` language roadmap would still likely have ecosystem connotations to people not familiar with the situation.

> I wonder if it's possible to have the PS language maintainers maintain two different roadmaps? If not, then the PS language maintainers would maintain a PS language roadmap, and a different place (I don't know who or where) would maintain a roadmap and version of the ecosystem.

This is an interesting idea. It could help coordinate effort around improving the PS ecosystem while the language further develops. Maybe even having both would help people see that PS has a `v1.0` language but not a `v1.0` ecosystem and alleviate some of gary's concerns?

Reply:
This action item for this thread depends on whether we want this document to include improvement suggestions.

Reply:
Agreed.

Reply:
In https://github.com/chexxor/purescript-documentation-discussion/pull/60/files#diff-3a877e274483d3781b26354f082146cc

I added the following point to the purpose:

> 4. Explore strategies for improving PureScript's documentation for its audiences.

If we accept that, then this document should include strategies/ideas for improvements. However, this merges together "Phase 2: defining the context/narrative" and "Phase 3: defining the purpose, principles, and outcome". This document is Phase 2, but this document also contains Phase 3, which is interesting...

Maybe we can revisit this separation after we merge all open PRs to view the latest version of this document as a whole. If we do that, then we should add this "two roadmaps" suggestion now and later decide whether to pull the strategies out of this document.

Reply:
The point of doing the Context/Narrative first is so that we do not prematurely define the Purpose, Principles, and Outcomes (PPO). It seems that, once one has defined the Context/Narrative, the desire to define PPO arises soon thereafter. The context provides a very nice flow into the PPO.

Speaking for myself, I've been finding it hard to not start thinking about solutions now that I have a better understanding of the problem.

This document also seems to be morphing from a "let's follow our methodology as much as possible" to "let's present our findings to people and actually start doing something to fix the problem."

> If we do that, then we should add this "two roadmaps" suggestion now and later decide whether to pull the strategies out of this document.

I had forgotten about that idea, but I agree that it should be included now. That seems to be the direction we're going. Also, this document seems to be fulfilling the purpose of that Discourse thread: answering how we can fix the situation.